File: services/qr_service.py
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

import pandas as pd
import qrcode
import io
import hashlib
from config import QR_DIR, PDF_DIR, BASE_URL
from utils.helpers import get_mexico_datetime

def save_excel():
    """Guarda el registro en formato excel"""
    from models.database import get_db
    import time
    db = get_db()
    
    # Obtener registros con reintentos
    max_retries = 3
    for attempt in range(max_retries):
        try:
            response = db.table('registros_asistencia').select('*').order('fecha_hora', desc=True).execute()
            registros_raw = response.data
            break
        except Exception as e:
            if attempt < max_retries - 1:
                time.sleep(0.5)
                continue
            else:
                raise Exception(f"Error de conexión con Supabase: {str(e)}")
    
    # Obtener todos los empleados y locales de una sola vez
    empleados_dict = {}
    locales_dict = {}
    
    try:
        emp_response = db.table('empleado').select('id, nombre').execute()
        empleados_dict = {emp['id']: emp['nombre'] for emp in emp_response.data}
        print(f"Empleados cargados: {len(empleados_dict)}")
    except Exception as e:
        print(f"Error al obtener empleados: {str(e)}")
    
    try:
        local_response = db.table('locales').select('id, nombre_local').execute()
        locales_dict = {loc['id']: loc['nombre_local'] for loc in local_response.data}
        print(f"Locales cargados: {len(locales_dict)}")
    except Exception as e:
        print(f"Error al obtener locales: {str(e)}")
    
    # Obtener datos relacionados usando los diccionarios
    registros = []
    for reg in registros_raw:
        empleado_nombre = empleados_dict.get(reg['empleado_id'], 'Desconocido')
        local_nombre = locales_dict.get(reg['locales_id'], 'Desconocido')
        
        registros.append({
            'ID': reg['id'],
            'Empleado': empleado_nombre,
            'Local': local_nombre,
            'Fecha': reg['fecha_hora'],
            'Observaciones': reg.get('observaciones', '')
        })

    # Crear DataFrame
    df = pd.DataFrame(registros)
    
    # Crear archivo Excel en memoria
    output = io.BytesIO()
    with pd.ExcelWriter(output, engine='openpyxl') as writer:
        df.to_excel(writer, index=False, sheet_name='Registros')
    output.seek(0)
    
    return output

def generar_pdf_qrs():
    """Generar PDF con códigos QR de locales"""
    try:
        from reportlab.lib.pagesizes import letter
        from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Image, PageBreak
        from reportlab.lib.styles import getSampleStyleSheet
    except Exception as e:
        print("No se pudo generar PDF. Instala reportlab:", e)
        return None

    from models.database import get_db
    PDF_DIR.mkdir(parents=True, exist_ok=True)
    pdf_path = PDF_DIR / "qrs_locales.pdf"

    db = get_db()
    response = db.table('locales').select('id, nombre_local').order('id').execute()
    locales = response.data

    doc = SimpleDocTemplate(str(pdf_path), pagesize=letter)
    styles = getSampleStyleSheet()
    story = []

    story.append(Paragraph("Códigos QR de Locales", styles["Title"]))
    story.append(Paragraph("Imprime estas hojas y pega cada QR en su local correspondiente.", styles["Normal"]))
    story.append(Spacer(1, 20))

    for i, local in enumerate(locales):
        qr_path = QR_DIR / f"local_{local['id']}.png"
        story.append(Paragraph(f"Local: {local['nombre_local']}", styles["Heading1"]))
        story.append(Spacer(1, 12))

        if qr_path.exists():
            story.append(Image(str(qr_path), width=260, height=260))

        if i < len(locales) - 1:
            story.append(PageBreak())

    doc.build(story)
    return pdf_path

def generate_qr_token(nombre_local, nombre_empleado, fecha, hora):
    """Generar token único para QR personalizado"""
    return hashlib.sha256(f"{nombre_local}{nombre_empleado}{fecha}{hora}{get_mexico_datetime().timestamp()}".encode()).hexdigest()[:12].upper()

def generate_qr_image(qr_url):
    """Generar imagen QR en base64"""
    import base64
    img = qrcode.make(qr_url)
    img_io = io.BytesIO()
    img.save(img_io, 'PNG')
    img_io.seek(0)
    return base64.b64encode(img_io.getvalue()).decode()

File: backend/src/services/qrService.ts
import QRCode from 'qrcode';
import { supabase } from '../config/database';
import * as XLSX from 'xlsx';
import { localRepository } from '../repositories/LocalRepository';
import { qrTokenRepository } from '../repositories/QrTokenRepository';
import { qrGeneradoRepository } from '../repositories/QrGeneradoRepository';

export interface QrTokenPayload {
  type: 'local_attendance';
  local_name: string;
  nombre_empleado?: string;
  fecha?: string;
  hora?: string;
  created_at?: string;
}

export function generateQrToken(nombre_local: string, nombre_empleado: string, fecha: string, hora: string): string {
  const payload: QrTokenPayload = {
    type: 'local_attendance',
    local_name: nombre_local,
    nombre_empleado,
    fecha,
    hora,
    created_at: new Date().toISOString()
  };

  return Buffer.from(JSON.stringify(payload)).toString('base64url');
}

export function decodeQrPayload(token: string): QrTokenPayload | null {
  if (!token) return null;

  try {
    const parsed = JSON.parse(Buffer.from(token, 'base64url').toString('utf8')) as Partial<QrTokenPayload>;
    if (parsed?.type === 'local_attendance' && typeof parsed.local_name === 'string' && parsed.local_name.trim()) {
      return parsed as QrTokenPayload;
    }

    return null;
  } catch {
    return null;
  }
}

export async function resolveLocalIdFromQrToken(token: string): Promise<string | null> {
  const payload = decodeQrPayload(token);

  if (payload?.local_name) {
    let locales = await localRepository.findByNombre(payload.local_name);
    if (locales.length === 0) {
      // Create local if it doesn't exist
      const newLocal = await localRepository.create({ nombre_local: payload.local_name });
      if (newLocal) {
        return newLocal.id;
      }
    } else {
      return locales[0].id;
    }
  }

  const qrTokenData = await qrTokenRepository.findByToken(token);
  if (qrTokenData?.local_id) {
    return qrTokenData.local_id;
  }

  const qrGenerado = await qrGeneradoRepository.findByToken(token);
  if (qrGenerado?.nombre_local) {
    let locales = await localRepository.findByNombre(qrGenerado.nombre_local);
    if (locales.length === 0) {
      // Create local if it doesn't exist
      const newLocal = await localRepository.create({ nombre_local: qrGenerado.nombre_local });
      if (newLocal) {
        return newLocal.id;
      }
    } else {
      return locales[0].id;
    }
  }

  return null;
}

export async function generateQrImage(qr_url: string): Promise<string> {
  try {
    const qrDataUrl = await QRCode.toDataURL(qr_url);
    return qrDataUrl;
  } catch (error) {
    console.error('Error generating QR image:', error);
    throw new Error('Error al generar imagen QR');
  }
}

export async function saveExcel(): Promise<Buffer> {
  try {
    // Get attendance records
    const { data: registros_raw, error } = await supabase
      .from('registros_asistencia')
      .select('*')
      .order('fecha_hora', { ascending: false });

    if (error) throw error;

    // Get employees and locales
    const { data: empleados } = await supabase
      .from('empleado')
      .select('id, nombre');
    
    const { data: locales } = await supabase
      .from('locales')
      .select('id, nombre_local');

    const empleados_dict = (empleados || []).reduce((acc, emp) => {
      acc[emp.id] = emp.nombre;
      return acc;
    }, {} as Record<string, string>);

    const locales_dict = (locales || []).reduce((acc, loc) => {
      acc[loc.id] = loc.nombre_local;
      return acc;
    }, {} as Record<string, string>);

    // Format records
    const registros = (registros_raw || []).map(reg => ({
      ID: reg.id,
      Empleado: empleados_dict[reg.empleado_id] || 'Desconocido',
      Local: locales_dict[reg.locales_id] || 'Desconocido',
      Fecha: reg.fecha_hora,
      Observaciones: reg.observaciones || ''
    }));

    // Create Excel workbook
    const worksheet = XLSX.utils.json_to_sheet(registros);
    const workbook = XLSX.utils.book_new();
    XLSX.utils.book_append_sheet(workbook, worksheet, 'Registros');

    // Write to buffer
    const excelBuffer = XLSX.write(workbook, { type: 'buffer', bookType: 'xlsx' });
    
    return excelBuffer as Buffer;
  } catch (error) {
    console.error('Error saving Excel:', error);
    throw new Error('Error al generar Excel');
  }
}
